src/utf8.rs

```diff
@@ -0,0 +1,506 @@
+{
+	let tag = tag_str.as_bytes();
+	Parser::new(move |input: &'a [u8], start: usize| {
+		let mut index = 0;
+		loop {
+			let pos = start + index;
+			if index == tag.len() {
+				let result = &input[start..pos];
+				// UNSAFE: Because slice is byte-identical to a str, it is known valid UTF-8
+				let result_str = unsafe { str::from_utf8_unchecked(result) };
+				return Ok((result_str, pos));
+			}
+			if let Some(s) = input.get(pos) {
+				if tag[index] != *s {
+					return Err(Error::Mismatch {
+						message: format!("seq {:?} at byte index: {}", tag, pos),
+						position: pos,
+					});
+				}
+			} else {
+				return Err(Error::Incomplete);
+			}
+			index += 1;
+		}
+	})
+}
+
+/// Success when current input symbol is one of the set.
+pub fn one_of<'a, S>(set: &'a S) -> Parser<'a, char>
+where
+	S: Set<char> + ?Sized,
+{
+	Parser::new(move |input: &'a [u8], start: usize| {
+		let (ch, size) = decode_utf8(&input[start..]);
+
+		if let Some(ch) = ch {
+			if set.contains(&ch) {
+				let pos = start+size;
+
+				Ok((ch, pos))
+			} else {
+				Err(Error::Mismatch {
+					message: format!("expect one of: {}, found: {}", set.to_str(), ch),
+					position: start,
+				})
+			}
+		} else {
+			no_utf8(start, size)
+		}
+	})
+}
+
+/// Success when current input symbol is none of the set.
+pub fn none_of<'a, S>(set: &'a S) -> Parser<'a, char>
+where
+	S: Set<char> + ?Sized,
+{
+	Parser::new(move |input: &'a [u8], start: usize| {
+		let (ch, size) = decode_utf8(&input[start..]);
+
+		if let Some(ch) = ch {
+			if !set.contains(&ch) {
+				let pos = start+size;
+
+				Ok((ch, pos))
+			} else {
+				Err(Error::Mismatch {
+					message: format!("expect one of: {}, found: {}", set.to_str(), ch),
+					position: start,
+				})
+			}
+		} else {
+			no_utf8(start, size)
+		}
+	})
+}
+
+/// Success when predicate returns true on current input symbol.
+pub fn is_a<'a, F>(predicate: F) -> Parser<'a, char>
+where
+	F: Fn(char) -> bool + 'a,
+{
+	Parser::new(move |input: &'a [u8], start: usize| {
+		let (ch, size) = decode_utf8(&input[start..]);
+
+		if let Some(ch) = ch {
+			if predicate(ch) {
+				let pos = start+size;
+
+				Ok((ch, pos))
+			} else {
+				Err(Error::Mismatch {
+					message: format!("is_a predicate failed on: {}", ch),
+					position: start,
+				})
+			}
+		} else {
+			no_utf8(start, size)
+		}
+	})
+}
+
+/// Success when predicate returns false on current input symbol.
+pub fn not_a<'a, F>(predicate: F) -> Parser<'a, char>
+where
+	F: Fn(char) -> bool + 'a,
+{
+	Parser::new(move |input: &'a [u8], start: usize| {
+		let (ch, size) = decode_utf8(&input[start..]);
+
+		if let Some(ch) = ch {
+			if !predicate(ch) {
+				let pos = start+size;
+
+				Ok((ch, pos))
+			} else {
+				Err(Error::Mismatch {
+					message: format!("is_a predicate failed on: {}", ch),
+					position: start,
+				})
+			}
+		} else {
+			no_utf8(start, size)
+		}
+	})
+}
+
+/// Read n chars.
+pub fn take<'a>(n: usize) -> Parser<'a, &'a str> {

```

Comment:
Notice: read n chars not n bytes, this is a difference from the base parser behavior